File: go/solutions/find_kth_smallest_pair_distance.go
package solutions

import "sort"

func SmallestDistancePair(nums []int, k int) int {
	sort.Ints(nums)
	low, high := 0, nums[len(nums)-1]-nums[0]
	countPairs := func(nums []int, distance int) int {
		count, left := 0, 0
		for right := 1; right < len(nums); right++ {
			for nums[right]-nums[left] > distance {
				left++
			}
			count += right - left
		}
		return count
	}
	for low < high {
		mid := low + (high-low)/2
		if countPairs(nums, mid) < k {
			low = mid + 1
		} else {
			high = mid
		}
	}

	return low
}

File: kotlin/src/main/kotlin/Find K-th Smallest Pair Distance.kt
package o.mg

fun smallestDistancePair(nums: IntArray, k: Int): Int {
    nums.sort()
    var low = 0
    var high = 1_000_000
    while (low <= high) {
        val mid = low + (high - low) / 2
        var j = 0
        if (k > nums.indices.sumOf { i ->
                while (nums[j] + mid < nums[i]) j++
                i - j
            }) low = mid + 1 else high = mid - 1
    }
    return low
}